### PR TITLE
Issue #7574 : Update doc for PackageAnnotation

### PR DESCRIPTION
**Pull request by PrabodDunuwila**
Issue : https://github.com/checkstyle/checkstyle/issues/7574
Screenshot from 2020-03-10 09-20-03

Configuration
```
$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
	<module name="PackageAnnotation" />
   </module>
</module>
```

Output with violation in MyClass.java
```
$ cat MyClass.java
@Deprecated
package com.example.annotations.packageannotation;

$ java -jar checkstyle-8.29-all.jar -c config.xml MyClass.java
Starting audit...
[ERROR] /home/dunu008/GSOC/CHECKSTYLE-PROJECT/tests/MyClass.java:2: Package annotations must be in the package-info.java info. [PackageAnnotation]
Audit done.
Checkstyle ends with 1 errors.
```
Fixing violation in MyClass.java
```
$ cat MyClass.java
package com.example.annotations.packageannotation;

$ java -jar checkstyle-8.29-all.jar -c config.xml MyClass.java
Starting audit...
Audit done.
```
Ouput in package-info.java
```
$ cat package-info.java
@Deprecated
package com.example.annotations.packageannotation;

$ java -jar checkstyle-8.29-all.jar -c config.xml package-info.java
Starting audit...
Audit done.
```

### CODE REVIEW

**Comment by rnveach** [comment]
@PrabodDunuwila Please see https://github.com/checkstyle/checkstyle/wiki/PR-rules .
Specifically items 3 and 4.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 1]
```diff
@@ -48,6 +48,14 @@
  * <pre>
  * &lt;module name=&quot;PackageAnnotation&quot;/&gt;
  * </pre>
+ * <p>
+ * Each package in a Java project should include a package-info.java file.
+ * In that file declare package annotations.
+ * </p>
+ * <pre>
+ * &#64;Deprecated
+ * package com.example.annotations.packageannotation;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: This read code of this Check to get to know when it violate package statement and when not. We need two examples: with violation (with trailing comment "violation") and without violation ( training comment "ok").

This Check is a bit unusual, as it depends on file name. So you need two code blocks with text above on what is name of file.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 2]
```diff
@@ -48,6 +48,14 @@
  * <pre>
  * &lt;module name=&quot;PackageAnnotation&quot;/&gt;
  * </pre>
+ * <p>
+ * Each package in a Java project should include a package-info.java file.
+ * In that file declare package annotations.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Please remove this paragraph

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 1 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Added file names which doesn't violates and violates naming convention.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 2 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Removed this paragraph.

**Comment by romani** [comment]
Please read PR rules and keep only one commit in PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: We need file name and their content. Your examples should show how Check works, it is based on file name and its content.

Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Changes done.Please review.

**Comment by PrabodDunuwila** [comment]
@romani please review

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 3]
```diff
@@ -48,6 +48,18 @@
  * <pre>
  * &lt;module name=&quot;PackageAnnotation&quot;/&gt;
  * </pre>
+ * <p>
+ * Examples of validating:
+ * </p>
+ * <pre>
+ * package-info.java     //ok, no violation
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: `ok, no violation` ==> `ok`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 4]
```diff
@@ -48,6 +48,18 @@
...
+ * &#64;Deprecated
+ * package com.example.annotations.packageannotation;
+ *
+ * package-information.java   //violation
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: package-information.java ==> MyClass.java

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 5]
```diff
@@ -48,6 +48,18 @@
  * <pre>
  * &lt;module name=&quot;PackageAnnotation&quot;/&gt;
  * </pre>
+ * <p>
+ * Examples of validating:
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: please move name of file out of "pre" to this line. (same for other example)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 3 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 4 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 5 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Please check whether I have addressed this point correctly.

**Comment by PrabodDunuwila** [comment]
@romani please review

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 6]
```diff
@@ -48,6 +48,18 @@
...
  * </pre>
+ * <p>
+ * Examples of validating files package-info.java and MyClass.java:
+ * </p>
+ * <pre>
+ * package-info.java     //ok
+ * &#64;Deprecated
+ * package com.example.annotations.packageannotation;
+ *
+ * MyClass.java   //violation
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: this line should be out of pre tag

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 7]
```diff
@@ -48,6 +48,18 @@
...
+ * &#64;Deprecated
+ * package com.example.annotations.packageannotation;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: this two lines should be in separate PRE tag

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 6 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 7 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

**Comment by rdiachenko** [comment]
> please regenerate image of how it looks like in web site.

@PrabodDunuwila please do, I still see an old image.

**Comment by PrabodDunuwila** [comment]
@rdiachenko updated the image.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 8]
```diff
@@ -48,6 +48,21 @@
  * <pre>
  * &lt;module name=&quot;PackageAnnotation&quot;/&gt;
  * </pre>
+ * <p>
+ * Examples of validating files package-info.java and MyClass.java:
+ * </p>
+ *
+ * <p>package-info.java     //ok</p>
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Why do you need those whitespaces between `package-info.java` and `//ok`?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 9]
```diff
@@ -48,6 +48,21 @@
...
+ * <pre>
+ * &#64;Deprecated
+ * package com.example.annotations.packageannotation;
+ * </pre>
+ *
+ * <p>MyClass.java   //violation</p>
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Same question about whitespaces as above.

File: src/xdocs/config_annotation.xml [hunk 10]
```diff
@@ -810,7 +810,17 @@ public static final int COUNTER = 10; // violation
 
       <subsection name="Examples" id="PackageAnnotation_Examples">
         <p> To configure the check:</p>
-        <source> &lt;module name=&quot;PackageAnnotation&quot;/&gt;
+        <source> &lt;module name=&quot;PackageAnnotation&quot;/&gt; </source>
+        <p>Examples of validating files package-info.java and MyClass.java:</p>
+        <p>package-info.java     //ok</p>
```
Comment (rdiachenko) [inline src/xdocs/config_annotation.xml]: Again, on your screenshot you have a single whitespace but here you use few of them.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 8 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Removed the white spaces.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 9 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Kept single white space.

File: src/xdocs/config_annotation.xml (hunk 10 above)
Comment (PrabodDunuwila) [reply src/xdocs/config_annotation.xml]: Kept single white space.

**Comment by PrabodDunuwila** [comment]
@rdiachenko I think now it makes more sense. Made the required changes.

**Comment by PrabodDunuwila** [comment]
@romani please review and merge.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 11]
```diff
@@ -48,6 +48,19 @@
...
+ * </p>
+ * <p>package-info.java</p>
+ * <pre>
+ * &#64;Deprecated
+ * package com.example.annotations.packageannotation; //ok
+ * </pre>
+ * <p>MyClass.java</p>
+ * <pre>
+ * &#64;Deprecated
+ * package com.example.annotations.packageannotation; //violation
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Please make violation case first and then show good case. This makes it look like we are changing bad code to good.

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Please sure another example of `MyClass` to show how to make violation go away (IE: remove annotation).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 12]
```diff
@@ -48,6 +48,19 @@
  * <pre>
  * &lt;module name=&quot;PackageAnnotation&quot;/&gt;
  * </pre>
+ * <p>
+ * Examples of validating files package-info.java and MyClass.java:
+ * </p>
+ * <p>package-info.java</p>
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Please add a little bit more than the file name so it is clear this is what you are referencing before the actual code.

I would combine `Examples of ...` with the file name for each example. Example: `Example of validating XXX file:`.
If some file is used more than once, say what makes the difference like "fixing violation".

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 11 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java (hunk 12 above)
Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

**Comment by PrabodDunuwila** [comment]
@rnveach please review

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java [hunk 13]
```diff
@@ -48,6 +48,20 @@
  * <pre>
  * &lt;module name=&quot;PackageAnnotation&quot;/&gt;
  * </pre>
+ * <p>Example of validating MyClass.java</p>
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: End line with semicolon to show the separation.
`Example of validating MyClass.java:`

Do this for the others too.

Comment (PrabodDunuwila) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/PackageAnnotationCheck.java]: Done

**Comment by rnveach** [comment]
Can be merged when CI passes if there are no more complaints

**Pull request merged**